### PR TITLE
Use Leibniz equality

### PR DESCRIPTION
**Pull request by hdgarrood**
Following on from #9

### CODE REVIEW

**Comment by hdgarrood** [comment]
cc @kl0tl

File: test/Main.purs [hunk 1]
```diff
@@ -15,5 +15,8 @@ instance newtypeRecordNewtype ::
   wrap = RecordNewtype <<< to
   unwrap (RecordNewtype rec) = from rec
 
+test2 :: forall ty row. TypeEquals row ( message :: String ) => Newtype ty (Record row) => ty -> String
+test2 = unwrap >>> typeEqualsProof >>> _.message
```
Comment (kl0tl) [inline test/Main.purs]: `message` seems like a good name for that function.

**Review by kl0tl — APPROVED** [review]
I would find an implementation of `from` based on `newtype From a b = From (b -> a)` a bit simpler:

```purs
from :: forall a b. TypeEquals a b => b -> a
from = case typeEqualsProof (From (\a -> a)) of From f -> f
```

especially if we don’t export `symm`.

A similar implementation of `to` based on a redundant `newtype To a b = To (a -> b)` would then be nice for symmetry.

This looks good to me regardless though!

File: src/Type/Equality.purs [hunk 2]
```diff
@@ -14,10 +15,22 @@ module Type.Equality
 -- | Note: any instance will necessarily overlap with
 -- | `refl` below, so instances of this class should
 -- | not be defined in libraries.
+class TypeEquals :: forall k. k -> k -> Constraint
 class TypeEquals a b | a -> b, b -> a where
-  to :: a -> b
-  from :: b -> a
+  typeEqualsProof :: forall p. p a -> p b
```
Comment (garyb) [inline src/Type/Equality.purs]: Maybe just `proof`?

Comment (hdgarrood) [reply src/Type/Equality.purs]: I think `proof` is more likely to be used as a local variable name, which is why I was thinking of giving this a longer name. I guess it depends whether we expect this module to be imported qualified or not?

Comment (garyb) [reply src/Type/Equality.purs]: I was imagining it'll be qualified - that's also the pattern we've been encouraging in most places - dropping type/class names from functions and members, like all the `Ref` stuff a while back. 

With it already having names like `to` and `from` I've only ever imported it qualified as it is.

Comment (hdgarrood) [reply src/Type/Equality.purs]: Yeah, true. It seems Type.Prelude currently re-exports the entirety of this module unqualified, which I think is perhaps not ideal. Come to think of it I'm not sure if people tend to import Type.Prelude qualified or not. Maybe we could just not re-export the new `proof` from Type.Prelude?

Comment (garyb) [reply src/Type/Equality.purs]: Not re-exporting `proof` sounds fine to me :+1:.

I'm not sure I've ever imported `Type.Prelude` at all, but if I were to do so I'd probably import it qualified, but then I import almost everything qualified these days, unless it's in our app prelude.

**Comment by hdgarrood** [comment]
If someone wants to carry this over the line I'd be grateful. I think all that's necessary is replacing `typeEqualsProof` with `proof`. I would also be happy to go with @kl0tl's suggestion of using newtypes `To` and `From` to implement `to` and `from`, or alternatively I think leaving the implementations of `to` and `from` as they are is probably fine too.

File: src/Type/Equality.purs (hunk 2 above)
Comment (kl0tl) [reply src/Type/Equality.purs]: Type.Prelude re-exports only TypeEquals, from and to so nothing to change there then o/

**Comment by kl0tl** [comment]
I pushed a few commits, this should be good to merge once Travis agrees.

**Pull request merged**